Multiple instances

jstandard opened the pull request:
@asteffey, here is the JSON feature for internal review.  Still working on XML.

jstandard committed:
Implements JSON aspect of feature request <Support ignoring order of array elements in JSON and XML>.  Added ignore-order array matcher to PactDsl and LambdaDsl on consumer side.  Added ignore-order matcher processing on provider side.

jstandard committed:
<Support ignoring order...> Update to provide different matching criteria for different instances of elements with the same name (including tests).

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt:
```diff
@@ -1,23 +1,7 @@
-import com.google.gson.JsonObject
-import com.google.gson.JsonPrimitive
+import au.com.dius.pact.core.model.matchingrules.*
+import com.google.gson.*
```
They seem to always use single name imports throughout their code.  Let's stay with their convention.

```suggestion
import au.com.dius.pact.core.model.matchingrules.DateMatcher
import au.com.dius.pact.core.model.matchingrules.IgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.IncludeMatcher
import au.com.dius.pact.core.model.matchingrules.MatchingRule
import au.com.dius.pact.core.model.matchingrules.MatchingRuleGroup
import au.com.dius.pact.core.model.matchingrules.MaxEqualsIgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MaxTypeMatcher
import au.com.dius.pact.core.model.matchingrules.MinEqualsIgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MinMaxEqualsIgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MinMaxTypeMatcher
import au.com.dius.pact.core.model.matchingrules.MinTypeMatcher
import au.com.dius.pact.core.model.matchingrules.NullMatcher
import au.com.dius.pact.core.model.matchingrules.NumberTypeMatcher
import au.com.dius.pact.core.model.matchingrules.RegexMatcher
import au.com.dius.pact.core.model.matchingrules.RuleLogic
import au.com.dius.pact.core.model.matchingrules.TimeMatcher
import au.com.dius.pact.core.model.matchingrules.TimestampMatcher
import au.com.dius.pact.core.model.matchingrules.TypeMatcher
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
```

asteffey replied:
In intellij, you can set this in preferences, code style, kotlin, imports (tab) and then set everything to single name.

asteffey committed:
wip: add more tests (failing)

Add more test cases, combining several into parameterized tests.

Add several test cases for extra elements.
Expect ignore-order, by itself, to expect arrays have equal length as this
is the default pact expectation.  To override this, ignore-order should
be combined with a mintype (or minmaxtype) matcher.

Several tests are failing. Include comments to explain why/rationale.

In some cases, added comments to discuss/propose more robust error messages.
Still unsure here, there's a trade between providing a useful explanation for failure
and just providing way too much info.

asteffey committed:
wip: another test

jstandard committed:
fix: Update JsonBodyMatching to handle
 * ignore-order with multiple of the same element
 * min & max ignore-order matching based on length of list
 * fixed several failing tests in JsonBodyMatchingSpec

jstandard commented:
@asteffey, Just pushed an update that fixes the tests you added earlier.  If this looks good, I'll still need to update the consumer side for building messages using Min/Max/MinMax-EqualIgnoreOrderMatcher.

Additionally, I'd like to meet and review tests, to ensure we've adequately tested all relevant scenarios.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"down"}]' | true  // expct[0]==actl[0] & expct[1]==actl[1]
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"up"}]'   | true  // expct[0]==actl[0] & expct[1]==actl[1]
+    '[{"id":"b", "status":"down"},{"id":"c", "status":"up"}]' | false // expct[0] no match & expct[1]==actl[1]
+// I don't feel this test should fail - let's discuss if you disagree
+//    '[{"id":"b", "status":"up"},{"id":"c", "status":"up"}]'   | false // expct[0|1]==actl[0], no unique for all expct
```
This should fail for the same reason the test on [L529](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L529) fails.

```suggestion
    '[{"id":"b", "status":"up"},{"id":"c", "status":"up"}]'   | false // expct[0|1]==actl[0], no unique for all expct
```

I also suggested adding a simpler test to show/test this in `'matching json bodies - returns a mismatch - with ignore-order - and multiple of the same element'`, showing that an actual of `'[100, 200]'` should match an expected of `'[100, 100]'` with `IgnoreOrderMatcher`.

jstandard replied:
I must not understand what you mean by 'no unique for all expected'

Expected [{"id":"a","status":"up"}, {"id":"b","status":"down"}]
Actual [{“id":"b", "status":"up"},{"id":"c", "status":"up"}]

$[0].id is RegexMatcher( a | b ) - check
$[0].status is EqualsMatcher( ) - check
$[1].id is RegexMatcher( b | c ) - check
$[1].status is RegexMatcher( up | down ) - check

asteffey replied:
Though expct[0] == actl[0] and expct[1] == actl[0], there is not a unique match for both expct[0] and expct[1].

This is just a more complicated case of expected=[100,100] and actual=[100,200].  Here, we also have expct[0] == actl[0] and expct[1] == actl[0], but because there are not two unique instances of 100 in actual we don't have a match.

jstandard replied:
Sorry, but I must still be missing something.  To me it looks like expct[0] == actl[0] and expct[1] == actl[1] given all the matching rules.  

I added the test for  expected=[100,100] and actual=[100,200] with just an IgnoreOrderMatcher (not MinEqualsIgnoreOrderMatcher) and that passed too.  Which I would've expected.

asteffey replied:
😅 oops, I was paying attention to the comment instead of the actual test case there...  You are :100: right, expct[0] == actl[0] and expct[1] == actl[1] for that test case (just as you have it for [L604](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L604) immediately below this comment.

Given that, I'd suggest changing the test case to this - which looks like it is currently passing.

```suggestion
    '[{"id":"b", "status":"up"},{"id":"a", "status":"down"}]' | false // expct[0]==expct[1]==actl[0], no unique for each
```

jstandard replied:
No problem.  Just want to make sure we are on the same page.  Better to go back and forth internally then find failures after we send out a merge request.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
+  }
+
+  def 'matching json bodies - return equality match - with ignore-order and min-equals matching'() {
+    given:
+    matchers.addCategory('body')
+      .addRule('$', new MinEqualsIgnoreOrderMatcher(1))
+
+    when:
+    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)
+
+    then:
+    mismatches.empty
+
+    where:
+    actualBody = OptionalBody.body('[200, 100, 300, "bad"]'.bytes)
+    expectedBody = OptionalBody.body('[100]'.bytes)
+  }
```
This test, as written, is really just tests the same functionality as [the test above it](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L420).

I'd recommend changing it back to what I was originally trying to get at, where the ignore-order and equals matchers may be passing the type matching still isn't.  That intent still isn't working.

```suggestion
  def 'matching json bodies - return type mismatch - with ignore-order and min-type matching'() {
    given:
    matchers.addCategory('body')
            .addRule('$', new MinEqualsIgnoreOrderMatcher(1))
            .addRule('$[0]', EqualsMatcher.INSTANCE)
            .addRule('$[*]', TypeMatcher.INSTANCE)

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    !mismatches.empty
    mismatches*.mismatch == ['Expected "bad" (JsonPrimitive) to be the same type as 100 (JsonPrimitive)']
    mismatches*.path == ['$.3']

    where:
    actualBody = OptionalBody.body('[200, 100, 300, "bad"]'.bytes)
    expectedBody = OptionalBody.body('[100]'.bytes)
  }
```

jstandard replied:
I've made several changes based on your suggestions, and **above** is the last test that is failing.  Also, I know why it's failing.  However, I'm not 100% sure how to fix it.  Although, I have an idea of a direction.

The inner for loop in compareListContentUnordered (**below**) compares each expectedValue to all actualValues, looking for a match.  Of course during the comparison it finds lots of mismatches which get ignored.  Only if there are no matches do we create a mismatch.

The problem is when ignoring mismatches that occur we eat 'Expected "bad" (JsonPrimitive) to be the same type as 100 (JsonPrimitive)', along with the equality mismatch.

Essentially in the for loop, some mismatches are valid and others aren't.  I can pretty easily create a data-structure that hangs on to the mismatches that occur.  The difficulty is knowing which mismatches are reportable and which aren't.  One course of action is to ignore equality mismatches and report the others?  

Let me know if you have any thoughts.  In the meantime, I'll start on the data-structure idea just to get a feel if this is a valid direction.

```
        for ((actualIndex, actualValue) in mutableActuals.withIndex()) {
          val outcome = compare(path + expectedIndex.toString(), expectedValue, actualValue, allowUnexpectedKeys, matchers)
          if (outcome.isEmpty()) {
            match = true                            // We've found a match
            mutableActuals.removeAt(actualIndex)    // remove it from list - there may be multiple of same in actual
            break                                   // move onto next item in expected
          } else {
            // do we need to keep track of mismatches?
            // sometimes there are valid mismatches that get consumed here
          }
        }
```

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/Matchers.kt:
```diff
@@ -3,15 +3,15 @@ package au.com.dius.pact.core.matchers
 import au.com.dius.pact.core.model.PathToken
-import au.com.dius.pact.core.model.matchingrules.MatchingRuleGroup
-import au.com.dius.pact.core.model.matchingrules.MatchingRules
+import au.com.dius.pact.core.model.matchingrules.*
```
They seem to always use single name imports throughout their code. Let's stay with their convention.

```suggestion
import au.com.dius.pact.core.model.matchingrules.IgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MatchingRuleGroup
import au.com.dius.pact.core.model.matchingrules.MatchingRules
import au.com.dius.pact.core.model.matchingrules.MaxEqualsIgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MinEqualsIgnoreOrderMatcher
import au.com.dius.pact.core.model.matchingrules.MinMaxEqualsIgnoreOrderMatcher
```
In intellij, you can set this in preferences, code style, kotlin, imports (tab) and then set everything to single name.

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/Matchers.kt:
```diff
@@ -74,6 +74,41 @@ object Matchers : KLogging() {
+   * TODO possibly remove this function
+   */
+  @JvmStatic
+  fun ignoreOrderMatchingEnabled(): Boolean = System.getProperty(PACT_MATCHING_IGNORE_ORDER)?.trim() == "true"
```
Agree with removing. Only place it is used is a test.
```suggestion
```

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/Matchers.kt:
```diff
@@ -3,15 +3,15 @@ package au.com.dius.pact.core.matchers
 
 object Matchers : KLogging() {
 
+  const val PACT_MATCHING_IGNORE_ORDER = "pact.matching.ignore-order"
```
Not needed if `ignoreOrderMatchingEnabled` is removed.

```suggestion
```

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/MatchersSpec.groovy:
```diff
@@ -64,6 +67,111 @@ class MatchersSpec extends Specification {
+    def testInvocation = { String value ->
+      System.setProperty('pact.matching.ignore-order', value)
+      Matchers.ignoreOrderMatchingEnabled()
+    }
```
Recommend removing along with `ignoreOrderMatchingEnabled` as this is the only place the method is used.

```suggestion
```

asteffey commented on core/model/src/main/kotlin/au/com/dius/pact/core/model/matchingrules/MatchingRules.kt:
```diff
@@ -224,6 +256,17 @@ data class MatchingRuleGroup @JvmOverloads constructor(
+              MaxEqualsIgnoreOrderMatcher(mapEntryToInt(map, MAX))
+            } else {
+              IgnoreOrderMatcher
+            }
```
This seems to work, but suggest adding a test cases to cover this in [MatchingRuleGroupSpec](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/model/src/test/groovy/au/com/dius/pact/core/model/matchingrules/MatchingRuleGroupSpec.groovy#L11) covering these cases:
```groovy
    [match: 'ignore-order']                 | IgnoreOrderMatcher             | 'if the matcher type is \'ignore-order\' and there is no min or max'
    [match: 'ignore-order', min: 1]         | MinEqualsIgnoreOrderMatcher    | 'if the matcher type is \'ignore-order\' and there is a min'
    [match: 'ignore-order', max: 1]         | MaxEqualsIgnoreOrderMatcher    | 'if the matcher type is \'ignore-order\' and there is a max'
    [match: 'ignore-order', max: 3, min: 2] | MinMaxEqualsIgnoreOrderMatcher | 'if the matcher type is \'ignore-order\' and there is a min and max'
```

asteffey commented on core/model/src/main/kotlin/au/com/dius/pact/core/model/matchingrules/MatchingRules.kt:
```diff
@@ -134,6 +134,38 @@ object ValuesMatcher : MatchingRule {
+/**
+ * Ignore order matching with a maximum size
+ */
+data class MaxEqualsIgnoreOrderMatcher(val max: Int) : MatchingRule {
```
I'm not sure why you mixed in equality into the Min/MaxIgnoreOrderMatchers?  Ignore order matching, regardless of whether or not it is combined with min or max, isn't unique to equality matching.  It should also work when combined with things like regex matching.

```suggestion
data class MaxIgnoreOrderMatcher(val max: Int) : MatchingRule {
```

jstandard replied:
As I understand it, the Min/Max/MinMaxTypeMatcher is only good for matching the size constraints *and* Type.  This is the reason why I couldn't wrap my head around why we would want to mix it with IgnoreOrderMatcher.  Since type doesn't care about the order.

I explicitly named them with Equals because they wouldn't work with some other type of matcher such as DateMatcher for example.

I don't know where that leaves us regarding the name change or perhaps the implementation of Min/Max/MinMax ignore order matchers.

asteffey replied:
But ignoreorder matchers (with or without the min/max) can be combined with any other type of matching, including type matching.

An example of this is in [`'matching json bodies - with ignore-order - and mixed elements'`](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L557) (the below snippet with the changes I had suggested previously).  Here, we want to do type matching on `4` in expected, but don't care in what position in the array a number shows up (and want to this not to match if a number isn't present):
```groovy
  @Unroll
  def 'matching json bodies - with ignore-order - and mixed elements'() {
    given:
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[*].id', new EqualsMatcher())
            .addRule('$[0].status', new EqualsMatcher())
            .addRule('$[1]', TypeMatcher.INSTANCE)
            .addRule('$[2].status', new RegexMatcher('up|down'))
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body('[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]'.bytes)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == matches

    where:
    actual                                                          | matches
    '[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]'   | true
    '[{"id":"a", "status":"up"}, "4", {"id":"b", "status":"down"}]' | false
    '[{"id":"b", "status":"up"}, {"id":"a", "status":"up"}, 5]'     | true
    '[{"id":"b", "status":"down"}, {"id":"a", "status":"up"}, 5]'   | true
    '[{"id":"b", "status":"up"}, {"id":"a", "status":"down"}, 5]'   | false
    '[{"id":"c", "status":"up"}, {"id":"a", "status":"up"}, 5]'     | false
  }
```

I had also suggested this other test case, which also provides an example of how this could work, where you still want to make sure the expected number is in the actual list (with an equality matcher, ignoring order) but also want to make sure that all the other items in the list are numbers (with a `$[*]` type matcher):
```groovy
  def 'matching json bodies - return type mismatch - with ignore-order and min-type matching'() {
    given:
    matchers.addCategory('body')
            .addRule('$', new MinEqualsIgnoreOrderMatcher(1))
            .addRule('$[0]', EqualsMatcher.INSTANCE)
            .addRule('$[*]', TypeMatcher.INSTANCE)
    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)
    then:
    !mismatches.empty
    mismatches*.mismatch == ['Expected "bad" (JsonPrimitive) to be the same type as 100 (JsonPrimitive)']
    mismatches*.path == ['$.3']
    where:
    actualBody = OptionalBody.body('[200, 100, 300, "bad"]'.bytes)
    expectedBody = OptionalBody.body('[100]'.bytes)
  }
```

The only special relationship I see between ignoreorder and equality is that ignoreorder matching should fall-back to basic equality matching if you apply that rule to a non-list, but it does this mostly as equality matching is the default form of matching in Pact.  But I'm not sure if that justifies it being in the name when weighed agains the fact that ignore-order matchers can be paired with other matchers...  If we do keep the Equals in the name here, we should also change `IgnoreOrderMatcher` to `EqualsIgnoreOrderMatcher` for consistency.

jstandard replied:
It all started with the test below.  However, we were trying to combine an IgnoreOrderMatcher with a MinTypeMatcher.  But since MinTypeMatcher only matches Types, all of the tests passed - even the third one which was supposed to fail.

There may have been another way to combine matchers and get the correct behavior.  I don't know.  However, we discussed this other MinIgnoreOrderMatcher - which became MinEqualsIgnoreOrderMatcher because the min/max/minmax matchers have another dimension to check against beyond just the array length (i.e. the individual elements).

I'll keep at this and think about it more.  We may want to setup a meeting to discuss further. 

```
@Unroll
  def 'matching json bodies - with ignore-order and min-equals matching - when actual has extra elements'() {
    given:
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body(expected.bytes)
    matchers.addCategory('body')
        .addRule('$', new MinEqualsIgnoreOrderMatcher(3))

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    mismatches.empty == match

    where:
    expected                | actual                            | match
    '[1,2]'                 | '[1,2,3]'                         | true
    '[1,2]'                 | '[2,1,3]'                         | true
    '[1,2]'                 | '[1,3,4]'                         | false
    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
  }

```

jstandard replied:
I've been thinking that it may be possible to pass the intended element matcher into Min/Max/MinMax IgnoreOrderMatcher along with the size(s).  Then rather than defaulting to Type or Equals, or whatever it could just use the matcher provided.

asteffey replied:
The `.addRule('$', new MinEqualsIgnoreOrderMatcher(3))` in that test was meant to replace the following [four rules](https://github.com/mitre/pact-jvm/blob/f37d290c63a3069196980b0bc91c296564e4045e/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L429):
```groovy
        .addRule('$', IgnoreOrderMatcher.INSTANCE)
        .addRule('$[0]', EqualsMatcher.INSTANCE)  // explicit EqualsMatcher needed to override MinTypeMatcher
        .addRule('$[1]', EqualsMatcher.INSTANCE)
        .addRule('$', new MinTypeMatcher(3))
```
In the original 4-rule version the MinTypeMatcher was used on the array itself to ensure that at least 3 elements were present in actual, but type matching shouldn't be applied to the elements because of the two explicit EqualsMatcher which override the type matching for those elements.  The IgnoreOrderMatcher allowed the 2 expected element values to be located out of order in the actual array.

Here is another version of the test with the ignore-order aspects commented out, for a point of reference. This all passes, including the test case with a false match.
```groovy
  @Unroll
  def 'matching json bodies - with ignore-order and min-type matching - when actual has extra elements'() {
    given:
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body(expected.bytes)
    matchers.addCategory('body')
            //.addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[0]', EqualsMatcher.INSTANCE)  // explicit EqualsMatcher needed to override MinTypeMatcher
            .addRule('$[1]', EqualsMatcher.INSTANCE)
            .addRule('$', new MinTypeMatcher(3))

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == match

    where:
    expected                | actual                            | match
    '[1,2]'                 | '[1,2,3]'                         | true
    //'[1,2]'                 | '[2,1,3]'                         | true
    '[1,2]'                 | '[1,3,4]'                         | false
    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
  }
```

The intent of MinIgnoreOrderMatcher was to just combine the Min feature of MinTypeMatcher with the IgnoreOrderMatcher, so that we could specify the need for an array which could be larger than the expected example without using a MinTypeMatcher and then overriding the type matching on the elements of the array.

I'm not sure I follow the need to pass another matcher into a Min/Max/MinMaxIgnoreOrderMatcher.  I think the current behavior of falling back to an equals matcher is correct (I had originally thought you should fallback to `matchIgnoringOrder` but later realized that `matchIgnoringOrder` should also just result in `matchEquals` in that situation).  If an element of the array needs to do something other than equals, you just define a matcher for that element (e.g. `$[0]`) or all elements (i.e. `$[*]`).

asteffey commented on core/model/src/main/kotlin/au/com/dius/pact/core/model/matchingrules/MatchingRules.kt:
```diff
@@ -134,6 +134,38 @@ object ValuesMatcher : MatchingRule {
+/**
+ * Ignore order matcher with a minimum size and maximum size
+ */
+data class MinMaxEqualsIgnoreOrderMatcher(val min: Int, val max: Int) : MatchingRule {
```
```suggestion
data class MinMaxIgnoreOrderMatcher(val min: Int, val max: Int) : MatchingRule {
```

asteffey commented on core/model/src/main/kotlin/au/com/dius/pact/core/model/matchingrules/MatchingRules.kt:
```diff
@@ -134,6 +134,38 @@ object ValuesMatcher : MatchingRule {
+/**
+ * Ignore order matcher with a minimum size
+ */
+data class MinEqualsIgnoreOrderMatcher(val min: Int) : MatchingRule {
```
```suggestion
data class MinIgnoreOrderMatcher(val min: Int) : MatchingRule {
```

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    where:
+    expected                | actual                            | match
+    '["a","A"]'             | '["b","B"]'                       | true
+    '["A","a"]'             | '["b","B"]'                       | true
```
The expected array should really always align with the matchers in-order.  It's the actual array order we want to vary.

```suggestion
    '["a","A"]'             | '["B","b"]'                       | true
```

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    def actualBody = OptionalBody.body(actual.bytes)
+    def expectedBody = OptionalBody.body(expected.bytes)
+    matchers.addCategory('body')
+        .addRule('$', IgnoreOrderMatcher.INSTANCE) // test passes when this is removed
```
```suggestion
        .addRule('$', IgnoreOrderMatcher.INSTANCE)
```

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    '[100, 100, 200, 200, 300]' | '[100, 300, 200, 100, 200, 400]' | true    | 5
+    '[100, 100, 200, 200, 300]' | '[100, 300, 300, 100, 200, 400]' | false   | 5 // only one 200 in actual
+  }
+
```
Some more suggested test cases to illustrate ignore order matching.  Currently, the equality version of the below test doesn't pass but the regex version of the test passes.

Oddly, the equality version does pass if you use a minignoreorder instead of ignoreorder, or if you explicitly add an equals matcher to `$[*]`.

```suggestion

def 'matching json bodies - returns a mismatch - with ignore-order - and multiple of the same element'() {
    given:
    def expected = '[100, 100]'
    def actual = '[100, 200]'
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            //.addRule('$', new MinEqualsIgnoreOrderMatcher(2)) // test passes when adding this?
            //.addRule('$[*]', EqualsMatcher.INSTANCE) // test also passes when adding this?

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    !mismatches.empty
    mismatches*.mismatch == ["Expected $expected to equal $actual ignoring order of elements"]
    mismatches*.path == ['$']
  }

  def 'matching json bodies - returns a mismatch - with ignore-order and regex - and multiple of the same element'() {
    given:
    def expected = '["red", "blue"]'
    def actual = '["blue", "seven"]'
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[*]', new RegexMatcher('red|blue'))

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    !mismatches.empty
    mismatches*.mismatch == ["Expected $expected to equal $actual ignoring order of elements"]
    mismatches*.path == ['$']
  }

```

jstandard replied:
The regex version you posted passes.  However, I don't feel it's equivalent.  Something like this would be equivalent...

    def expected = '["red", "red"]'
    def actual = '["red", "blue"]'
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)

which actually doesn't return any mismatches, but it should.

asteffey replied:
I agree this version is more equivalent, but it really isn't exercising any execution paths that aren't already covered elsewhere.  We already have test coverage trying out ignore-order with different data types (number, string, etc).

Testing with a regex matcher tests that ignore-order works with more than just default equality matching, and the control logic for dealing with those was significantly changed with the code changes here.

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt:
```diff
@@ -119,10 +105,51 @@ fun <M : Mismatch> domatch(
+    emptyList()
+  } else {
+    listOf(mismatchFactory.create(expected, actual, "Expected ${valueOf(expected)} to equal ${valueOf(actual)} ignoring order of elements", path))
+  }
```
This seems duplicative of [`compareListContentUnordered`](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/JsonBodyMatcher.kt#L108).

When I tried to change it to the below, more tests seemed to pass (with the exception changes to the error messages in [`'matching json bodies - return a mismatch - with ignore-order - when actual is missing an element'`](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L366)).


```suggestion
  return when {
    actual is JsonArray && expected is JsonArray -> emptyList()
    else -> matchEquality(path, expected, actual, mismatchFactory)
  }
```

In addition to JsonArray, we would probably also need to cover cases like List and Element.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+            .addRule('$', IgnoreOrderMatcher.INSTANCE)
+            .addRule('$[*].id', new EqualsMatcher())
+            .addRule('$[0].status', new EqualsMatcher())
+            .addRule('$[1].status', new RegexMatcher('up|down'))
+            .addRule('$[3]', TypeMatcher.INSTANCE)
```
This test should have been failing, but is passing.  The rules were applying to the wrong elements in expected. There is no `$[3]` and `$[1]` is `4` so there is no `$[1].status`...

That test should not have been passing as written - I'm not sure why they were.

```suggestion
            .addRule('$[1]', TypeMatcher.INSTANCE)
            .addRule('$[2].status', new RegexMatcher('up|down'))
```

jstandard replied:
Yes, I see that it was passing with the rules defined incorrectly.  I'm not sure why that is.  I haven't touched anything regarding how the rules are applied.  However, I should look into it further.

I made the changes you suggested, and the tests pass.

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/JsonBodyMatcher.kt:
```diff
@@ -132,6 +175,22 @@ object JsonBodyMatcher : BodyMatcher, KLogging() {
+            path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))
+        }
+      }
+      result
```
This is a complicated flow...  It seems that the last `else` is really just covering the case where `Matchers.oneOfIgnoreOrderMatchersDefined("body", path, matchers) == true`, so would suggest reordering.

jstandard replied:
I can see your reasoning for reordering this.  My only hesitation is that it follows the same order as it did previously and just tacks ignore-order onto the end.  I could make ignore-order first like so...

```
return if (Matchers.isEqualsIgnoreOrderMatcherDefined("body", matchers)) {
    // match unordered list
} else if (Matchers.matcherDefined("body", path, matchers)) {
  // match ordered list - at least one body matcher defined
} else {
  // match ordered list - no body matcher, default to equals matcher
}
```

jstandard replied:
Actually that order failed on the test with one ordered and one unordered...

```
  def 'matching json bodies - return a mismatches - with ignore-order set on another list'() {
    given:
    matchers.addCategory('body')
      .addRule('$[0].array1', EqualsIgnoreOrderMatcher.INSTANCE)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).find {
      it instanceof BodyMismatch && it.mismatch.contains('Expected 1 but received 2')
    }

    where:
    actualBody = OptionalBody.body('''{
     "array1": [{"foo": "a"},{"foo": "b"}],
     "array2": [2, 3, 1, 4]
     }'''.bytes)
    expectedBody = OptionalBody.body('''{
     "array1": [{"foo": "b"},{"foo": "a"}],
     "array2": [1, 2, 3, 4]
     }'''.bytes)
  }
```

asteffey replied:
Does `Matchers.isEqualsIgnoreOrderMatcherDefined` not take a path, or did you just omit it from your code snippet?  It would need to the path, especially in a situation like this test.

jstandard replied:
Well it did, but it wasn't being used, so I removed it.  When I tried this, I thought about adding it back and testing using matchesPath like wildcardMatcherDefined does.  

However, I don't think that would work because it filters on path.size which is the same for both array1 and array2 in this case.

```
  @JvmStatic
  fun wildcardMatcherDefined(path: List<String>, category: String, matchers: MatchingRules?) =
    if (matchers != null) {
      val resolvedMatchers = matchers.rulesForCategory(category).filter(Predicate {
        matchesPath(it, path) == path.size
      })
      resolvedMatchers.matchingRules.keys.any { entry -> entry.endsWith(".*") }
    } else false
```

jstandard replied:
I'll give it shot tho.

jstandard replied:
So, I made the change to filter on path length and that broke some tests in JsonBodyMatcherSpec.  

Originally I was looking at 'isEqualsIgnoreOrderMatcherDefined' and 'isOtherThanEqualsIgnoreOrderMatcherDefined' as being not path based.  Now I'm wondering if they should be based on the path.  However, the length alone is not specific enough because that will generate a lot of false positives.

I'm going to have to think about this a little more.

jstandard replied:
I was able to fix this.  Turned out well, because it identified another couple of problems that would have cropped up later.

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/JsonBodyMatcher.kt:
```diff
@@ -132,6 +175,22 @@ object JsonBodyMatcher : BodyMatcher, KLogging() {
+          result.add(BodyMismatch(a, b,
+            "Expected a List with ${expectedValues.size()} elements but received ${actualValues.size()} elements",
+            path.joinToString("."), generateJsonDiff(expectedValues, actualValues)))
+        }
```
This doesn't seem right.  I would think that this would be covered by [matchMinEquals](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt#L420) and [matchMaxEquals](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt#L451).

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt:
```diff
@@ -390,6 +417,68 @@ fun <M : Mismatch> matchMaxType(
   }
 }
 
+fun <M : Mismatch> matchMinEquals(
```
would matchMinIgnoreOrder be a better name?

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt:
```diff
@@ -390,6 +417,68 @@ fun <M : Mismatch> matchMaxType(
+      emptyList()
+    }
+  } else {
+    matchEquality(path, expected, actual, mismatchFactory)
```
shouldn't this call matchIgnoringOrder?

asteffey replied:
Actually, this is probably the right behavior - especially if you change the implementation of `matchIgnoringOrder` as I had suggested above.  Calling matchIgnoringOrder would just duplicate the checking of if actual is a JsonArray/Element/List - which is unnecessary.

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/MatcherExecutor.kt:
```diff
@@ -119,10 +105,51 @@ fun <M : Mismatch> domatch(
             matchMaxType(matcher.max, path, expected, actual, mismatchFn)
     is IncludeMatcher -> matchInclude(matcher.value, path, expected, actual, mismatchFn)
     is NullMatcher -> matchNull(path, actual, mismatchFn)
+    is IgnoreOrderMatcher -> matchIgnoringOrder(path, expected, actual, mismatchFn)
+    is MinEqualsIgnoreOrderMatcher -> matchMinEquals(matcher.min, path, expected, actual, mismatchFn)
+    is MaxEqualsIgnoreOrderMatcher -> matchMaxEquals(matcher.max, path, expected, actual, mismatchFn)
+    is MinMaxEqualsIgnoreOrderMatcher -> matchMinEquals(matcher.min, path, expected, actual, mismatchFn) +
+            matchMaxEquals(matcher.max, path, expected, actual, mismatchFn)
```
wouldn't you want/need to try the more specific cases after the less specific case?

```suggestion
    is MinMaxEqualsIgnoreOrderMatcher -> matchMinEquals(matcher.min, path, expected, actual, mismatchFn) +
            matchMaxEquals(matcher.max, path, expected, actual, mismatchFn)
        is MinEqualsIgnoreOrderMatcher -> matchMinEquals(matcher.min, path, expected, actual, mismatchFn)
    is MaxEqualsIgnoreOrderMatcher -> matchMaxEquals(matcher.max, path, expected, actual, mismatchFn)
    is IgnoreOrderMatcher -> matchIgnoringOrder(path, expected, actual, mismatchFn)
```

jstandard replied:
I can change this.  I was just following the form used for MinType, ...

asteffey replied:
Looking again, the order shouldn't matter...   I'm not sure why I was thinking that did.

jstandard replied:
No, the order doesn't matter.  I thought you wanted the change for aesthetics.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+    '[1, "b", 3, "d"]'                   | '["d", 1, 3, "b"]'
+    '[{"i": "a"}, {"i": 2}, {"i": "c"}]' | '[{"i": 2}, {"i": "c"}, {"i": "a"}]'
+  }
+
```
I was able to think of a few more test cases for situations with lists in lists (which should or should not inherit the ignore-order matching).  There are probably some more corner cases to identify, but these first few seem to work.

```suggestion

def 'matching json bodies - return no mismatches - with ignore-order on nested lists'() {
    given:
    def expected = '[[1,2,3],[4,5,6]]';
    def actual = '[[6,4,5],[2,3,1]]';
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty
  }

  def 'matching json bodies - return no mismatches - with ignore-order and overriding equality'() {
    given:
    def expected = '[[1,2,3],[4,5,6]]';
    def actual = '[[6,4,5],[2,3,1]]';
    def expectedBody = OptionalBody.body(expected.bytes)
    def actualBody = OptionalBody.body(actual.bytes)
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[0]', EqualsMatcher.INSTANCE)

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    !mismatches.empty
    mismatches*.mismatch = ['Expected [[1,2,3], [4,5,6]] to equal [[6,4,5], [2,3,1]] ignoring order of elements']
    mismatches*.path = ['$']
  }
```

jstandard replied:
The second test fails miserably with 13 mismatches.  Likely from this rule:
 .addRule('$[0]', EqualsMatcher.INSTANCE)

jstandard replied:
If I change it to:  '.addRule('$[0].*', EqualsMatcher.INSTANCE)' it generates no mismatches, which I think makes sense.  

I understand what you'd like to test, but think that extra code block we added negatively impacted this test case.

jstandard replied:
The second test will not pass because .addRule('$[0]', EqualsMatcher.INSTANCE) gets applied recursively and generates 13 mismatches (not just the one listed).  I've modified the test to just check !mismatches.empty.

asteffey reviewed (changes requested):
I've identified a few more test cases, or really variants of the tests already in place, which aren't passing.

I also provided some feedback/suggestions to the code base.  Let me know if you have any questions and we can talk through them.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+
+    where:
+    actual                                                        | matches
+    '[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]' | true
```
I'd recommend also adding in this test case:
```suggestion
    '[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]' | true
    '[{"id":"a", "status":"up"}, "4", {"id":"b", "status":"down"}]' | false
```

jstandard committed:
fix: more updates to JsonBodyMatching

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/util/CollectionUtils.kt:
```diff
@@ -29,4 +29,4 @@ fun <E> List<E>.padTo(size: Int, item: E): List<E> {
     return list
   }
-}
+}
```
```suggestion
}

```

New line existed in original version, https://github.com/mitre/pact-jvm/blob/3bc07b834f651f1aa17663416786cdae7a9ce0d6/core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/util/CollectionUtils.kt.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/util/CollectionUtilsSpec.groovy:
```diff
@@ -37,5 +37,4 @@ class CollectionUtilsSpec extends Specification {
     [1, 2, 3] | 6    | [1, 2, 3, 1, 1, 1]
   }
-
 }
```
```suggestion

}
```

Blank line existed in original.  https://github.com/mitre/pact-jvm/blob/3bc07b834f651f1aa17663416786cdae7a9ce0d6/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/util/CollectionUtilsSpec.groovy#L40

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +330,335 @@ class JsonBodyMatcherSpec extends Specification {
+    '[{"id":"b", "status":"up"},{"id":"a", "status":"down"}]' | false // expct[0]==expct[1]==actl[0], no unique for each
+    '[{"id":"b", "status":"up"},{"id":"c", "status":"up"}]'   | true  // expct[0]==actl[0] & expct[1]==actl[1]
+    '[{"id":"c", "status":"up"},{"id":"b", "status":"up"}]'   | true  // expct[0]==actl[1] & expct[1]==actl[0]
+  }
```
```suggestion
    '[{"id":"b", "status":"up"},{"id":"a", "status":"up"}]'   | true // expct[0]==actl[1] & expct[1]==actl[0]
  }
```

Fails because expct[0] also == actl[0], which breaks current algorihtm in compareListContentUnordered

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy:
```diff
@@ -328,4 +330,335 @@ class JsonBodyMatcherSpec extends Specification {
+    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
+  }
+
+  @Unroll
+  def 'matching json bodies - with ignore-order and regex - return no mismatches'() {
+    given:
+    def actualBody = OptionalBody.body(actual.bytes)
+    def expectedBody = OptionalBody.body(expected.bytes)
+    matchers.addCategory('body')
+      .addRule('$', EqualsIgnoreOrderMatcher.INSTANCE)
+      .addRule('$[0]', new RegexMatcher('[a-z]'))
+      .addRule('$[1]', new RegexMatcher('[A-Z]'))
+
+    expect:
+    matcher.matchBody(expectedBody, actualBody, true, matchers).empty
+
+    where:
+    expected                | actual                            | match
+    '["a","A"]'             | '["b","B"]'                       | true
+    '["a","A"]'             | '["B","b"]'                       | true
+  }
```
Small edit to make use of `match` and add a negative case.

```suggestion
@Unroll
  def 'matching json bodies - with ignore-order and regex - return no mismatches'() {
    given:
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body(expected.bytes)
    matchers.addCategory('body')
      .addRule('$', EqualsIgnoreOrderMatcher.INSTANCE)
      .addRule('$[0]', new RegexMatcher('[a-z]'))
      .addRule('$[1]', new RegexMatcher('[A-Z]'))

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == matches

    where:
    expected    | actual      || matches
    '["a","A"]' | '["b","B"]' || true
    '["a","A"]' | '["B","b"]' || true
    '["a","A"]' | '["B","B"]' || false
  }
```

asteffey committed:
EqualsIgnoreOrderMatcher expects list of same size

Change to EqualsIgnoreOrderMatcher so it expects actual list
to be same size as the expected list, making it more consistent with
regular EqualsMatcher.

Min/MaxEqualsIgnoreOrderMatcher can be used when an actual list
of a different size is expected.

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/MatchersSpec.groovy:
```diff
@@ -64,6 +68,88 @@ class MatchersSpec extends Specification {
 
+  def '[*] matcher defined - should be true when star matcher on array'() {
+    expect:
+    Matchers.isArrayStarMatcherDefined(['$', 'any'], 'body', matchingRules())
```
`['$', 'any']` would be a path for an object, like `{"any": "value"}`, but we're talking about arrays, so should be using indexes.  Test fails when we use an index.

```suggestion
    Matchers.isArrayStarMatcherDefined(['$', '0'], 'body', matchingRules())
```

asteffey commented on core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/MatchersSpec.groovy:
```diff
@@ -64,6 +68,88 @@ class MatchersSpec extends Specification {
+      matchingRules
+    }
+  }
+
```
This test also fails, but should pass.

```suggestion

  def '[*] matcher defined - should be false'() {
    expect:
    !Matchers.isArrayStarMatcherDefined(['$', '2'], 'body', matchingRules())

    where:
    matchingRules = {
      def matchingRules = new MatchingRulesImpl()
      matchingRules.addCategory('body')
        .addRule('$', new MinEqualsIgnoreOrderMatcher(3))
        .addRule('$[0]', TypeMatcher.INSTANCE)
        .addRule('$[1]', TypeMatcher.INSTANCE)
      matchingRules
    }
  }
```

asteffey commented on core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/Matchers.kt:
```diff
@@ -74,6 +78,34 @@ object Matchers : KLogging() {
+  fun isArrayStarMatcherDefined(path: List<String>, category: String, matchers: MatchingRules?) =
+    if (matchers != null) {
+      val resolvedMatchers = matchers.rulesForCategory(category).filter(Predicate {
+        matchesPath(it, path) == 0
```
`matchesPath` is a little confusing, it could use a KDoc.

It returns what index (starting from 1) that a matching expression matches against a path array.  Here are some examples I ran to clarify for myself:
```groovy
    Matchers.INSTANCE.matchesPath('$.a', ['$', 'a']) == 2
    Matchers.INSTANCE.matchesPath('$.a', ['$', 'a', 'b']) == 2
    Matchers.INSTANCE.matchesPath('$.a.b', ['$', 'a', 'b']) == 3
    Matchers.INSTANCE.matchesPath('$.a.c', ['$', 'a', 'b']) == 0 // no match
    Matchers.INSTANCE.matchesPath('$.a.0', ['$', 'a', '0']) == 3
    Matchers.INSTANCE.matchesPath('$.a.0', ['$', 'a', '1']) == 0 // no match
    Matchers.INSTANCE.matchesPath('$.a[*]', ['$', 'a', '0']) == 3
    Matchers.INSTANCE.matchesPath('$.*', ['$', 'a', '0']) == 2
    Matchers.INSTANCE.matchesPath('$[*]', ['$', 'a', '0']) == 0 // because a isn't array index
    Matchers.INSTANCE.matchesPath('$[*]', ['$', '0', '0']) == 2
```

Given this, I don't think using matchesPath would work.  I'm still not sure about a way that would work, other than trying to get matchers for an out-of-bound index.

asteffey reviewed (changes requested):
In addition, I also pushed https://github.com/mitre/pact-jvm/commit/1ae49c2c78647efbb2b8ebe009915860be739638 which changes
EqualsIgnoreOrderMatcher so it expects the actual list
to be same size as expected, making it more consistent with
regular EqualsMatcher.

jstandard committed:
fix: removed unused test parameters

jstandard committed:
feat: Consumer side DSL updates for ignore-order matching

jstandard commented:
@asteffey, Is this ready to merge?  I pushed the consumer side updates.  I don't think I can create a pull request until this is merged.  Maybe we can discuss next week.
